lib/spack/spack/directives.py

```diff
@@ -585,6 +588,9 @@ def depends_on(
     @see The section "Dependency specs" in the Spack Packaging Guide.
 
     """
+    if spack.spec.Spec(spec).name in SUPPORTED_LANGUAGES:
+        assert type == "build", "languages must be of 'build' type"
```

Comment:
Why is this needed?  It's not really a build dependency; it's a build dependency on the compiler *and* (maybe) a link dependency on the runtime, depending on what the compiler decides to do. 

Languages should just omit the deptype. In practice, it doesn't matter b/c it's determined by the compiler's package (and some may not need a runtime). The caller of `depends_on` should not have to do anything based on *how* the language is implemented. they should just say they depend on it.

Reply:
> Why is this needed? It's not really a build dependency; **it's a build dependency on the compiler** and (maybe) a link dependency on the runtime, depending on what the compiler decides to do.

Depending on a language just brings in the compiler, so as you say above it's a build dependency on the compiler.  Then the compiler may or may not decide to inject further dependencies. If we don't make the language dependency a build dependency we won't be able to have duplicates (since link dependency virtuals are unified, whereas build are not).

> Languages should just omit the deptype.

But `depends_on` have that argument, so a user may specify it. That was one of the reason for which I started implementing this using another directive. If we use `depends_on` I think people need to get used to adding `type="build"` as they do e.g. for `cmake`. If we use a wrapper directive we can take control of the type, and avoid exposing it.